docs/proposals/manifest-hydrator.md

```diff
@@ -0,0 +1,538 @@
+
+Equipped with this information, the developer can perform the exact same steps as Argo CD and be confident that their dry manifest changes will produce the desired output.
+
+Ensuring that hydration is deterministic assures the developer that the output for a given dry state will be the same next week as it is today.
+
+###### Avoiding Esoteric Behavior
+
+We should avoid the developer needing to know Argo CD-specific behavior in order to reproduce hydration. Tools like Helm, Kustimize, etc. have excellent public-facing documentation which the developer should be able to take advantage of without needing to know quirks of Argo CD.
+
+##### Reliable Previews
+
+Deterministic hydration output allows Argo CD to produce a reliable change preview when a developer proposes a change to the dry manifests via a PR.
+
+If output is not deterministic, then a preview generated today might not be valid/correct a week, day, or even hour later. Non-determinism makes it so that developers can't trust that the change they review will be the change actually applied.
+
+### Non-Goals
+
+1) Implementing a change promotion system
+
+## Open Questions
+
+* The `sourceHydrator` field is mutually exclusive with the `source` and the `sources` field. Should we throw an error if they're both configured, or should we just pick one and ignore the others?
+* How will/should this feature relate to the image updater? Is there an opportunity to share code, since both tools involve pushing to git?
+* Should we enforce a naming convention for hydrated manifest branches, e.g. `argo/...`? This would make it easier to recommend branch protection rules, for example, only allow pushes to `argo/*` from the argo bot.
+* Should we enforce setting a `sourceHydrator.syncSource.path` to something besides `.`? Setting a path makes it easier to add/remove other apps later if desired.
+
+## Proposal
+
+Today, Argo CD watches one or more git repositories (configured in the `spec.source` or `spec.sources` field). When a new commit appears, Argo CD updates the desired state by rendering the manifests with the configured manifest hydration tool. If auto-sync is enabled, Argo CD applies the new manifests to the cluster.
+
+With the introduction of this change, Argo CD will watch two revisions in the same git repository: the first is the "dry source", i.e. the git repo/revision where the un-rendered manifests reside, and the second is the "hydrated source," where the rendered manifests are places and retrieved for syncing to the cluster.
+
+### New `spec.sourceHydrator` Application Field
+
+A `sourceHydrator` field will be added to the Argo CD Application spec:
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: example
+spec:
+  # The sourceHydrator field is mutually-exclusive with `source` and with `sources`. If this field is configured, we 
+  # should either throw an error or ignore the other two.
+  sourceHydrator:
```

Comment:
Putting aside the question of whether or not this should be a separate project vs. in-tree for now, is there a strong reason do the hydration fields need to be introduced _into the Application spec itself_ as opposed to a CRD that references the Application?

In today's contributor meeting, there was mention of field duplication as one reason and the problem of keeping the values in-sync, but I feel this would not be an issue if a hydration CRD could simply reference the Application spec so there is still single source of truth (the App spec.source).

Second, in the demo, it was shown that in order to use the feature, a user would have to rename `source` to `sourceHydrator`, then add `drySource` + `syncSource` to the spec. This means there's no easy migration path for users since they would then need to regenerate their application specs to turn this on. A separate CRD solves this since this feature can be turned on and off more easily. In other words, they would create the CR to turn it on, delete the CR to turn it off, without touching the original Application.

Another concern I have is that this feature will inevitably need more and more status/bookkeeping about hydration, which will then spill into the Application status itself. So in addition to health status, sync status, then there will need to be hydration status and all the details there.

I just feel that the concerns of hydration (writing) to a repo and syncing (reading) from a repo have a natural clean separation, both logically as well as technically. So I feel we can & should preserve that separation as separate CRDs.

Reply:
I think embedding sourceHydrator in the Application CR has some crucial advantages:

1. UX: It provides individual developers with easy/intuitive source hydration via existing and familiar interfaces
2. Architecture: It's consistent with Argo CD’s existing CR design
3. Maintenance: It takes advantage of Argo CD’s existing API/UI code and permissions structure

I think there are two intuitive options for implementing this as a separate CRD: A) SourceHydrator that references an Application or B) an Omnibus SourceHydrator that handles multiple apps. I'll answer both options.

### A) SourceHydrator references Application

This option would probably look like this:

```yaml
kind: SourceHydrator
spec:
  application: my-app
  targetRevision: main
  path: envs/dev/west
---
kind: Application
spec:
  source:
    repoURL:
    targetRevision: env/dev
    path: west
```

#### 1. Impacts on UX

1. This requires maintaining an additional resource, which could be a pain for end users
2. The SourceHydrator spec isn't super intuitive, because you don't know at a glance the repoURL, targetRevision, path, or project; you have to inspect two resources
3. The Application `source` field becomes confusing, because not all features are actually available. 
   * `targetRevision`: normally you can use a commit tag, hash, or semver range as the target revision. But the hydrator as currently designed cannot push to a commit tag, and it fundamentally cannot push to a hash or semver range (those would be nonsensical destinations, technically allowed by the spec).
   * `chart`, `helm`, `kustomize`, and `directory.jsonnet`: none of these fields are meaningful when using rendered manifests, and using some will throw runtime errors. None of those restrictions will be clear by looking at the spec.
   * `plugin` and `directory`: using either of these fields defeats much of the point of using rendered manifests, because it applies last-mile transformations that aren't recorded to git. This downside won't be immediately apparent to the user.

#### 2. Impacts on Architecture

Having a "flat config" is consistent with how Argo CD has built most of its features. If there's a 1:1 relationship between some configuration and the Application CR, then we've embedded that configuration (multi-source support is a good example).

This design is in contrast with how Flux has done things. By breaking parts of their API into separate CRDs, they've simplified the mental model from an implementation perspective. But I think that users have appreciated Argo's approach of "you write one file, and you're done."

Traditionally in Argo CD if there is some configuration which is "about" an application, it is stored in the Application CR. We believe that individual users do and should care about manifest hydration. Our experience with the rendered manifest pattern at Intuit is that if the hydration process is "hidden" from the user, it results in confusion and frustration.

Having a separate CRD for Application-specific config would be possible. But it would be a departure from the pattern that's been very successful so far.

#### 3. Impacts on Maintenance

Using a separate CRD would require new API/UI/CLI code tailored to handling the new CR. For example, if we want to let users update the SourceHydrator config in the UI, we'll need either a new API to handle those updates or to expand the existing API to handle the additional resource. It's more code and complexity to maintain and reason about. All these are big reasons ApplicationSets still lacks a robust UI/CLI experience.

Security controls would become more difficult to think about. For example, if we want to use the Application's AppProject to restrict destination repositories and allowed app namespaces, we'd need code to pull the Application config, pull it's AppProject config, and then apply the rules. In a single CR, we could just call the existing functions to enforce those rules. To enforce apps-in-any-namespace restrictions, the app controller currently has code to filter out apps in disallowed namespaces. If we used a separate CRD, we'd have to reimplement the same logic in a different controller and either duplicate the restrictions config (two env vars) or write additional code to pull the restrictions from some common config location. 

### B) Omnibus SourceHydrator

Another option would be to write a CRD that handles hydration completely independently of applications, maybe even hydrating for multiple applications based on patterns. For example:

```yaml
kind: SourceHydrator
spec:
  repoURL: https://git.example.com/org/repo
  targetRevision: main
  targetBranches:
  - name: env/dev
    paths:
    - source: envs/dev/west
      destination: west
    - source: envs/dev/east
      destination: east
# etc.
```

This has some advantages. For example, you can look in one place to understand how all the hydration works for an app.

But you still have all the same downsides as described above.

You also have the problem that this type of omnibus config is difficult for individual users to deal with. As a new resource without a 1:1 app relationship, we'd need a new API and RBAC to provide a strong UI experience (similar to AppSets). In practice, it will probably be maintained by some ops team and become opaque to the developer. We believe the developer should fully understand and be in full control of hydration, because dry manifests are their interface, and hydrated manifests are what gets applied to the cluster. The user cares about both.

This resource also just obscures the inherent 1:1 relationship. One hydrated manifest always has one hydrated source. By listing them all in one file, you just add one layer of obfuscation between the user and the hydration process.

### Answers to specific concerns

> there's no easy migration path for users since they would then need to regenerate their application specs to turn this on

The Application source has to be modified either way. If the SourceHydrator is configured separately, the user still has to change the `spec.source` field to point to the hydrated instead of the DRY source. They'd also need to remove incompatible fields like `kustomize`, `helm`, etc. (if they know that they need to do so, since it's not obvious from the spec).

> this feature will inevitably need more and more status/bookkeeping about hydration, which will then spill into the Application status itself

It will, but since hydration is "about" the Application, this is a reasonable place to do that bookkeeping. I’m not sure what the ultimate concern is here. Storing information about the app is what the status field is for.

> the concerns of hydration (writing) to a repo and syncing (reading) from a repo have a natural clean separation, both logically as well as technically

They have a logical and clean separation in the same way that “pulling from a repo” and “syncing to Kubernetes” do: they’re logically separate, but they’re both config “about” the Application. The fact that they're very separate tasks makes it easy to write code to handle the two parts separately, just like we do for “pulling” and “syncing” today.

Reply:
I agree that the source hydrator should be part of the `Application` spec rather than another CRD for most of the reasons outlined, plus one more: references create further lifecycle concerns that need to be taken care of. For example, do you delete the hydrator or the application first? Do you put a finalizer on one or the other? Is one _owned_ by the other? While these aren't difficult problems to solve and there's prior art within the project, it's one more thing to consider for both maintainers and users.

Reply:
I support including the source hydrator in the `Application` spec instead of creating a separate CRD, primarily for the reasons outlined above. 

Additionally, our current `Application` spec already contains a similar scenario with our Multi-Source features `sources` field. In multi-source applications, we disregard the `source` field when `sources` is provided in the action spec. This rule is something we follow across UI / CLI / API.

We could apply the same logic with `sourceHydrator`—for instance, if `sourceHydrator` is specified, we could either ignore the `source` and `sources` fields or trigger an error, if necessary.